### DISCUSSION TITLE
Why does unicode rendering randomly fail?

### DISCUSSION THREAD

**Discussion by the-solipsist** [kovidgoyal/kitty#6718] [Q&A]
Once in a while, in some tabs of kitty, display of unicode characters fails.  But at the same time, it works fine on other windows even within the same tab.  If I open a new tab / a new instance of kitty, it works fine.

In this screenshot, the `â¹` is supposed to be `₹`.
![image](https://github.com/kovidgoyal/kitty/assets/3080824/6774ac66-ae3f-4d48-abc7-08d28f3ae05f)

In this screenshot, I'm guessing that the ` No such interface âorg.freedesktop.portal.Settingsââ` is supposed to be `No such interface “org.freedesktop.portal.Settings”`
![Screenshot from 2023-10-16 02-24-39](https://github.com/kovidgoyal/kitty/assets/3080824/73be5c8b-7ef1-445f-96b1-341071abce50)

Any ideas what the issue could be?

**Comment by kovidgoyal** [comment]
It will be because of different fallback fonts. Your different kitty instances will have seen fallback characters in different orders, leading to different fonts being used. kitty first looks up existing fallback fonts before querying fontconfig for a fallback. So If character x1 causes fallback font f1 to be loaded, and then kitty sees character x2 it will first check if f1 has x2 and use it if so. However if it sees x2 before x1 it will query fontconfig for x2 which might give a font different than f1.

Use --debug-font-fallback and find out which fonts on your system have the corrupted glyphs and remove them and you will be fine.

**Reply by the-solipsist** [reply]
I don't see how "different fallback fonts" can be the answer, since each kitty instance is using the same font (NotoSansMono-Regular.ttf) which has support for symbols like `“` and `€` and `₹`.

**Comment by the-solipsist** [comment]
I can reliably replicate this. 

Each time I open and close either [`hledger-iadd`](https://github.com/hpdeifel/hledger-iadd) or [`hledger-ui`](https://hackage.haskell.org/package/hledger-ui), Unicode symbols like `₹` stop displaying in that window.  I'm not sure what about opening those Haskell-based TUI software has this impact, nor how to go about debugging this further.

**Comment by kovidgoyal** [comment]
You were told what causes it and how to debug it further. If you choose not to believe what you were told, that's your look out..
